### ISSUE TITLE
Sometimes bounding box coordinates are correct, but the mask is either missing or only covers a very small part of the detected object

### ISSUE THREAD

**Issue by synthdatagit** [ultralytics/ultralytics#16733]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Similar to "closed" bug report: https://github.com/ultralytics/ultralytics/issues/7704

### Environment

Ultralytics 8.3.0 🚀 Python-3.9.18 torch-1.13.0+cu116 CUDA:0 (NVIDIA A100-SXM4-40GB, 40444MiB)
YOLO11x-seg summary (fused): 491 layers, 62,007,903 parameters, 0 gradients, 318.5 GFLOPs

### Minimal Reproducible Example

image_src = cv2.imread(img_path)

results = model.predict(
    source=image_src, save=False, conf=0.4, imgsz=cfg['imgsz']
)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @synthdatagit, thank you for bringing this to our attention 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com) for insights, particularly if you're new to the platform, where Python and CLI usage examples might help tackle common issues.

As you're experiencing a 🐛 Bug, we appreciate the minimal reproducible example you've provided. This will greatly assist in diagnosing the problem.

For now, please ensure:

- Your environment is up-to-date. Upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) setup with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

  ```bash
  pip install -U ultralytics
  ```

- YOLO can be run in several environments, such as [Google Colab](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb) or through [Docker](https://hub.docker.com/r/ultralytics/ultralytics) 📦, that are fully configured.

For community and discussion:
- Join the [Discord](https://ultralytics.com/discord) 🎧 for live chat.
- Visit our [Discourse](https://community.ultralytics.com) or [Subreddit](https://reddit.com/r/ultralytics) for more detailed conversations.

✅ [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests, visible on this badge, should be green indicating all functionalities are operational:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Please note this is an automated response. An Ultralytics engineer will review your issue shortly. Thank you for your patience! 😊

**Comment by Y-T-G** [comment]
Do you have images?

**Comment by Y-T-G** [comment]
You can train with `overlap_mask=False` if you want to avoid the masks being removed due to redundant boxes.

https://docs.ultralytics.com/modes/train/#train-settings

**Comment by synthdatagit** [comment]
> You can train with `overlap_mask=False` if you want to avoid the masks being removed due to redundant boxes.
> 
> https://docs.ultralytics.com/modes/train/#train-settings

Okay, I'll try that.

**Comment by glenn-jocher** [comment]
@synthdatagit great! Let us know if you have any further questions.

**Comment by synthdatagit** [comment]
> @synthdatagit great! Let us know if you have any further questions.

Setting overlap_mask = False did not solve the issue.

Here is my current training call:

model = YOLO(cfg["model"])  # cfg["model"] = 'yolo11x-seg.pt'

results = model.train(
        data=cfg["data"],
        cache=False,
        epochs=cfg["epochs"],
        batch=cfg["batch"],
        optimizer=cfg["optimizer"],
        lr0=cfg["lr0"],
        plots=True,
        save=True,
        imgsz=cfg["imgsz"],
        project=project,
        name=name,
        workers=cfg["workers"],
        pretrained=cfg["pretrained"],
        device=cfg["device"],
        overlap_mask=cfg['mask_overlap'], # which is set to False
        exist_ok=False,
        val=True,
        classes = cfg["classes"]
    )

**Comment by Y-T-G** [comment]
Provide images of the issue.

**Comment by yuriarfil** [comment]
I am facing this problem. Both bounding box and mask are well defined in prediction but it's missing when re-draw the result.masks.xy with cv2.drawContours.  
I've tried this solution also "train with overlap_mask=False" but did not solve. 
Did I miss something?
![yolov8mask](https://github.com/user-attachments/assets/82ac00db-a94d-4113-96b1-42367605dffa)

**Comment by yuriarfil** [comment]
@Y-T-G  Could you please show me how to do that

**Comment by Y-T-G** [comment]
You can try adding this before your inference code.

```
from functools import partial
import ultralytics.utils.ops as ops
import ultralytics.engine.results as results

# Create a new partial function with strategy="concat" by default
new_masks2segments = partial(ops.masks2segments, strategy="concat")

# Monkey patch the original method
ops.masks2segments = new_masks2segments
results.ops.masks2segments = new_masks2segments
```

**Comment by synthdatagit** [comment]
> You can try adding this before your inference code.
> 
> ```python
> from functools import partial
> import ultralytics.utils.ops as ops
> import ultralytics.engine.results as res
> 
> # Create a new partial function with strategy="concat" by default
> new_masks2segments = partial(ops.masks2segments, strategy="concat")
> 
> # Monkey patch the original method
> ops.masks2segments = new_masks2segments
> res.ops.masks2segments = new_masks2segments
> ```

Hi, that seems to solve my issue.

What caused the problem? And how was it solved by this?

Thanks

**Comment by Y-T-G** [comment]
If your mask for an object is split into multiple segments, then by default, `ultralytics` will only save the largest part.

**Comment by synthdatagit** [comment]
> If your mask for an object is split into multiple segments, then by default, `ultralytics` will only save the largest part. The code changes it to save all the parts.

Okay, is there a way to use the same arguments for model.val ?

**Comment by synthdatagit** [comment]
> The same code should work for val

The validation plots I get are still missing the masks. And the F1 score for the mask (for the object class that is the issue) is still very low. The F1 box score for the same class is high

**Comment by synthdatagit** [comment]
# Create a new partial function with strategy="concat" by default
    new_masks2segments = partial(ops.masks2segments, strategy="concat")

    # Monkey patch the original method
    ops.masks2segments = new_masks2segments
    res.ops.masks2segments = new_masks2segments
    
    metrics = model.val(
        data=cfg["data"],
        cache=False,
        plots=True,
        save=True,
        save_json=True,
        imgsz=cfg["imgsz"],
        project=project,
        name=name,
        device=cfg["device"],
        exist_ok=False,
        split='test',
        classes = cfg["classes"]
    )

**Comment by Y-T-G** [comment]
> The validation plots I get are still missing the masks.

Did you run `model.val()` after adding that code?

> And the F1 score for the mask (for the object class that is the issue) is still very low. The F1 box score for the same class is high

The score calculation wouldn't be affected by this. It only affects the plot and returned value from `xy` and `xyn` method.

**Comment by Y-T-G** [comment]
> The validation plots I get are still missing the masks.

The plot also wouldn't be affected by this.

**Comment by yuriarfil** [comment]
> You can try adding this before your inference code.
> 
> ```python
> from functools import partial
> import ultralytics.utils.ops as ops
> import ultralytics.engine.results as res
> 
> # Create a new partial function with strategy="concat" by default
> new_masks2segments = partial(ops.masks2segments, strategy="concat")
> 
> # Monkey patch the original method
> ops.masks2segments = new_masks2segments
> res.ops.masks2segments = new_masks2segments
> ```

Thanks @Y-T-G it's getting better but still have missing mask!
![yolov8mask-1](https://github.com/user-attachments/assets/5956d739-3a9d-4757-a171-0f314e94e78b)

**Comment by Y-T-G** [comment]
@yuriarfil You can use the masks directly.

```python
for masks in results[0].masks.data.cpu().numpy():
     ...
```

**Comment by synthdatagit** [comment]
> > The validation plots I get are still missing the masks.
> 
> The plot also wouldn't be affected by this.

Okay, but can you then explain why the F1 score for boxes is good and the F1 score for masks is bad.

This observation corresponds perfectly with the fact that masks are not generated correctly on the specific object, but the bounding box is.

The change of mask2segment strategy worked when I used the model.predict call, which I use for my own evaluation methods (where the F1 score is now higher for the masks) but it fails when model.val is used.

**Comment by synthdatagit** [comment]
> @yuriarfil You can use the masks directly.
> 
> ```python
> for masks in results[0].masks.data.cpu().numpy():
>      ...
> ```

Without changing the strategy to 'concat' this numpy array would not cover the entire object. I have tried this

**Comment by synthdatagit** [comment]
> > @yuriarfil You can use the masks directly.
> > ```python
> > for masks in results[0].masks.data.cpu().numpy():
> >      ...
> > ```
> 
> Without changing the strategy to 'concat' this numpy array would not cover the entire object. I have tried this

Sorry, what I have tried is:

xy = results[0].masks.xy
xyn = results[0].masks.xyn

and then drawing the polygons with OpenCV

**Comment by Y-T-G** [comment]
The metrics are calculated using the masks directly, and the plots are also created using the masks directly. That's why `result.plot()` doesn't have missing masks. If the score is low, it's probably due to some other reason. The `.xy` or `.xyn` segments are not used for score calculation or plotting.

**Comment by synthdatagit** [comment]
> The metrics are calculated using the masks directly, and the plots are also created using the masks directly. That's why `result.plot()` doesn't have missing masks. If the score is low, it's probably due to some other reason. The `.xy` or `.xyn` segments are not used for score calculation or plotting.

But the validation plots (generated by model.val) has missing masks.

**Comment by Y-T-G** [comment]
Does `model.predict()` show the missing mask for those same images?

**Comment by yuriarfil** [comment]
> @yuriarfil You can use the masks directly.
> 
> ```python
> for masks in results[0].masks.data.cpu().numpy():
>      ...
> ```

Just try but still has missing mask
![yolov8mask-2](https://github.com/user-attachments/assets/9c57aafe-ac5e-4ff7-a251-6adb915b79e0)

**Comment by Y-T-G** [comment]
@yuriarfil You can't use masks in `drawContours`. You have to use it in `addWeighted` or first run `findContours` on mask and then draw the contours.

**Comment by synthdatagit** [comment]
> Does `model.predict()` show the missing mask for those same images?

When I use model.predict after the change of strategy to 'concat' then the 

xy = results[0].masks.xy
xyn = results[0].masks.xyn

has more points = better mask on the specific object.

I have used these polygons in my own evaluation. Hence I get a better F1 score after applying the 'concat' strategy. However this is not the case with the built-in model.val plots.

So why is this?

**Comment by Y-T-G** [comment]
I will need to see the images of the same object from the same image missing a mask in the `model.val()` plots but having a mask in `result.plot()`.

**Comment by yuriarfil** [comment]
> @yuriarfil You can't use masks in `drawContours`. You have to use it in `addWeighted` or first run `findContours` on mask and then draw the contours.

It works, thanks @Y-T-G 
![yolov8mask-3](https://github.com/user-attachments/assets/171d0c48-4cf3-4fdf-931c-3ccfcbf7b554)

**Comment by dstanner** [comment]
I am still getting polygons with zero elements and empty masks, even with the patch described above.

```python
import numpy as np

from functools import partial
import ultralytics.utils.ops as ops
import ultralytics.engine.results as res
from ultralytics import YOLO

# add the patch at the top of my inference script
new_masks2segments = partial(ops.masks2segments, strategy="concat")
ops.masks2segments = new_masks2segments
res.ops.masks2segments = new_masks2segments

model = YOLO(
    # filename
)

pred = model.predict(
    # image file,
    conf=0.15,
)
pred = pred[0].to('cpu')

polygons = [x.xy for x in pred.masks]

for ind, x in enumerate(polygons):
    if x[0].shape[0] == 0:
        print(ind, x[0].shape)
            # 43 (0, 2)
            # 47 (0, 2)
            # 49 (0, 2)
```

So, out of 58 detections in this prediction, 3 of them had segmentation masks with zero elements. The sum of the prediction masks for these three detections was also 0 (zero pixels marked positive). The boxes for these predictions were correct, but the masks still missing. 

Anything else I need to do to stop the model from returning empty segmentation masks when there is a corresponding box detected?

**Comment by Y-T-G** [comment]
@dstanner Does it have mask when you draw them using `result.plot()`?

**Comment by dstanner** [comment]
@Y-T-G There are bounding boxes for the missing mask detections in the plot, but I don't see masks in those bounding boxes. 

I should say that most of the images I'm working with have dozens or even hundreds of objects, many of which are small -- maybe 10 x 10 pixels.

The missing masks in the test image I'm working with are very small. Could it be that small objects can be detected, but with no mask generated?

Two very small crops of an example image here, with the boxes with missing masks circled.

![example](https://github.com/user-attachments/assets/91934553-9db7-44dd-b61c-fb7a9fdc3c1b)

![example2](https://github.com/user-attachments/assets/af9f1552-d18e-433f-bf07-ac2f449349dc)

**Comment by Y-T-G** [comment]
@Laughing-q Are boxes returned if there are no associated masks? There's also another issue with the same problem.

https://github.com/ultralytics/ultralytics/issues/16925
